### PR TITLE
[storybook__addon-info] Update types to match 5.2

### PR DESCRIPTION
**Pull request by gaetanmaisse**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/storybookjs/storybook/tree/next/addons/info (I followed README and source code to update the types accordingly)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

--

What I did? 

 - Use types directly from `Storybook` repo as it is now written in TS
 - Update types to match version 5.2 of the addon and of Storybook (as they have to have the same version)

### CODE REVIEW

File: types/storybook__addon-info/index.d.ts [hunk 1]
```diff
@@ -1,50 +1,68 @@
...
-      description: string;
-      defaultValue: any;
-    }>
-  }>;
-  excludedPropTypes?: string[];
+    text?: string;
+    header?: boolean;
+    inline?: boolean;
+    source?: boolean;
+    propTables?: Array<ComponentType<any>> | false;
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/index.d.ts]: Use `unknown` instead? 🤔

```suggestion
    propTables?: Array<ComponentType<unknown>> | false;
```

Or just delete the `any`? 🤔
```suggestion
    propTables?: ComponentType[] | false;
```

File: types/storybook__addon-info/index.d.ts [hunk 2]
```diff
@@ -1,50 +1,68 @@
...
+    propTablesExclude?: Array<ComponentType<any>>;
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/index.d.ts]: Use `unknown` instead? 🤔

```suggestion
    propTablesExclude?: Array<ComponentType<unknown>>;
```

Or just delete the `any`? 🤔
```suggestion
    propTablesExclude?: ComponentType[];
```

File: types/storybook__addon-info/index.d.ts [hunk 3]
```diff
@@ -1,50 +1,68 @@
...
+    styles?: object;
+    components?: { [key: string]: ComponentType<any> };
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/index.d.ts]: Use `unknown` instead? 🤔

```suggestion
    components?: { [key: string]: ComponentType<unknown> };
```

Or just delete the `any`? 🤔
```suggestion
    components?: { [key: string]: ComponentType };
```

File: types/storybook__addon-info/index.d.ts [hunk 4]
```diff
@@ -1,50 +1,68 @@
...
+    maxPropObjectKeys?: number;
+    maxPropArrayLength?: number;
+    maxPropStringLength?: number;
+    TableComponent?: ComponentType<{
+        propDefinitions: Array<{
+            property: string;
+            propType: object | string; // TODO: info about what this object is...
+            required: boolean;
+            description: string;
+            defaultValue: any;
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/index.d.ts]: Use `unknown` instead? 🤔

```suggestion
            defaultValue: unknown;
```

File: types/storybook__addon-info/storybook__addon-info-tests.tsx [hunk 5]
```diff
@@ -1,77 +1,119 @@
...
 const { Component } = React;
 
-const TableComponent = ({ propDefinitions }: { propDefinitions: Array<{
-    property: string;
-    propType: { [key: string]: any} | string;
-    required: boolean;
-    description: string;
-    defaultValue: any;
-}> }) => (
+const TableComponent = ({
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/storybook__addon-info-tests.tsx]: Make this a `FunctionComponent` instead

```ts
type Props = ComponentProps<Options['TableComponent']>;
const TableComponent : FunctionComponent<Props> = ({ propDefinitions }) => (
  // ...
);
```

File: types/storybook__addon-info/storybook__addon-info-tests.tsx [hunk 6]
```diff
@@ -1,77 +1,119 @@
-/// <reference types="storybook__react" />
-
 import * as React from 'react';
-import { addDecorator, storiesOf } from '@storybook/react';
+import { addDecorator, storiesOf, addParameters } from '@storybook/react';
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/storybook__addon-info-tests.tsx]:
```suggestion
import { addDecorator, addParameters, Options, storiesOf } from '@storybook/react';
```

File: types/storybook__addon-info/storybook__addon-info-tests.tsx [hunk 7]
```diff
@@ -1,77 +1,119 @@
-/// <reference types="storybook__react" />
-
 import * as React from 'react';
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/storybook__addon-info-tests.tsx]:
```suggestion
import React, { Component, ComponentProps } from 'react';
```

File: types/storybook__addon-info/storybook__addon-info-tests.tsx [hunk 8]
```diff
@@ -1,77 +1,119 @@
...
-import { addDecorator, storiesOf } from '@storybook/react';
+import { addDecorator, storiesOf, addParameters } from '@storybook/react';
 import { setDefaults, withInfo } from '@storybook/addon-info';
 
 const { Component } = React;
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/storybook__addon-info-tests.tsx]:
```suggestion
```

File: types/storybook__addon-info/storybook__addon-info-tests.tsx [hunk 9]
```diff
@@ -1,77 +1,119 @@
-/// <reference types="storybook__react" />
-
 import * as React from 'react';
-import { addDecorator, storiesOf } from '@storybook/react';
+import { addDecorator, storiesOf, addParameters } from '@storybook/react';
 import { setDefaults, withInfo } from '@storybook/addon-info';
```
Comment (MichaelDeBoey) [inline types/storybook__addon-info/storybook__addon-info-tests.tsx]:
```suggestion
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by gaetanmaisse** [comment]
Thanks for the review @MichaelDeBoey! I already tried to use `unknown` but it's causing TS issues in "tests" because there are more props in the given component than in `ComponentType<unknown>` 🤷‍♂ I'm just a React newbie and don't know a workaround.

I created an interface for the props of `TableComponent` used in `Options` of this addon and was able to use it directly in tests then 😃

**Comment by MichaelDeBoey** [comment]
@gaetanmaisse I would just delete the `any` between the `ComponentType` props then, since the `props` type is defaulted to `{}`.

**Comment by gaetanmaisse** [comment]
@MichaelDeBoey we don't want it to be defaulted to `{}` as given component can have required props. 
I added a 2nd commit with a more realistic use of the addon parameters, in this example `label` is a required prop in both `BaseButton` and `CustomButton` component so I have to keep `ComponentType<any>` in the types to avoid:
```
TS2322: Type 'FunctionComponent<HTMLAttributes<HTMLButtonElement> & { label: string; }>[]' is not assignable to type 'ComponentType<{}>[]'.
  Type 'FunctionComponent<HTMLAttributes<HTMLButtonElement> & { label: string; }>' is not assignable to type 'ComponentType<{}>'.
    Type 'FunctionComponent<HTMLAttributes<HTMLButtonElement> & { label: string; }>' is not assignable to type 'FunctionComponent<{}>'.
      Types of parameters 'props' and 'props' are incompatible.
        Type '{ children?: ReactNode; }' is not assignable to type 'PropsWithChildren<HTMLAttributes<HTMLButtonElement> & { label: string; }>'.
          Property 'label' is missing in type '{ children?: ReactNode; }' but required in type '{ label: string; }'.

```

**Comment by MichaelDeBoey** [comment]
Now there's no example with `TableComponent` anymore? 🤔

**Comment by gaetanmaisse** [comment]
`TableComponent` is now only used as it was before (and as it is in addon readme and SB examples) so as 
`TableComponent` property of addon Options:  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40270/files#diff-507fc2a1a9fafe89f249889154266a38R58 😉

**Review by MichaelDeBoey — APPROVED** [review]

**Review by ndelangen — APPROVED** [review]

**Review by MichaelDeBoey — APPROVED** [review]

**Pull request merged**